packages/nextra/src/server/rehype-plugins/twoslash/rehype-twoslash-popup.ts

```diff
@@ -0,0 +1,133 @@
+import { rendererRich, type RendererRichOptions } from '@shikijs/twoslash'
+import type { Element, ElementContent, Root } from 'hast'
+import type { Code } from 'mdast'
+import { fromMarkdown } from 'mdast-util-from-markdown'
+import { gfmFromMarkdown } from 'mdast-util-gfm'
+import { defaultHandlers, toHast } from 'mdast-util-to-hast'
+import type { ShikiTransformerContextCommon } from 'shiki'
+import type { Plugin } from 'unified'
+import { EXIT, visit } from 'unist-util-visit'
+
+const TWOSLASH_POPUP_IMPORT_AST = {
+  type: 'mdxjsEsm',
+  data: {
+    estree: {
+      body: [
+        {
+          type: 'ImportDeclaration',
+          source: { type: 'Literal', value: 'nextra/components' },
+          specifiers: ['Popup', 'PopupTrigger', 'PopupPanel'].map(name => ({
+            type: 'ImportSpecifier',
+            imported: { type: 'Identifier', name },
+            local: { type: 'Identifier', name }
+          }))
+        }
+      ]
+    }
+  }
+} as any
+
+export const rehypeTwoslashPopup: Plugin<[], Root> = () => ast => {
+  // The tagName is being converted to lowercase when calling the shiki.codeToHtml method inside rehypePrettyCode.
+  // convert it back to Uppercase.
+  visit(ast, { tagName: 'popup' }, node => {
+    node.tagName = 'Popup'
+  })
+  visit(ast, { tagName: 'popuptrigger' }, node => {
+    node.tagName = 'PopupTrigger'
+  })
+  visit(ast, { tagName: 'popuppanel' }, node => {
+    node.tagName = 'PopupPanel'
+  })
+
+  visit(ast, { tagName: 'code' }, node => {
+    if (node.data?.meta === 'twoslash') {
+      ast.children.unshift(TWOSLASH_POPUP_IMPORT_AST)
+      return EXIT
+    }
+  })
+}
+
+export function twoslashRenderer(options?: RendererRichOptions) {
+  return rendererRich({
+    ...options,
+    renderMarkdown,
+    renderMarkdownInline,
+    hast: {
+      hoverToken: {
+        tagName: 'Popup'
+      },
+      hoverPopup: {
+        tagName: 'PopupPanel'
+      },
+      hoverCompose: ({ popup, token }) => [
+        popup,
+        {
+          type: 'element',
+          tagName: 'PopupTrigger',
+          properties: {},
+          children: [
+            {
+              type: 'element',
+              tagName: 'span',
+              properties: {
+                class: 'twoslash-hover'
+              },
+              children: [token]
+            }
```

Comment:
this is not needed 
<img width="490" alt="image" src="https://github.com/user-attachments/assets/571a34b1-81fc-469d-8ec4-83d9e4c8a0f2">